### ISSUE TITLE
Improve GPU-STUMP Speed

### ISSUE THREAD

**Issue by seanlaw** [TDAmeritrade/stumpy#245]
[This follows from a discussion on the Numba discourse channel](https://numba.discourse.group/t/understanding-cuda-stream/216)

Currently, STUMPY can be executed on both the CPUs and the GPUs (i.e., given the same input, both should produce the same results - this is useful for validation). The GPU kernel is essentially self contained in the [`gpu_stump.py`](https://github.com/TDAmeritrade/stumpy/blob/master/stumpy/gpu_stump.py) module. In fact, all of the heavy lifting occurs in [repeatedly launching a single GPU kernel](https://github.com/TDAmeritrade/stumpy/blob/164368e97e2acae5f882a3acb5e7caefc144b7a0/stumpy/gpu_stump.py#L343-L362) called [_compute_and_update_PI_kernel](https://github.com/TDAmeritrade/stumpy/blob/164368e97e2acae5f882a3acb5e7caefc144b7a0/stumpy/gpu_stump.py#L17-L168):

```
        for i in range(range_start, range_stop):
            _compute_and_update_PI_kernel[blocks_per_grid, threads_per_block](
                i,
                device_T_A,
                device_T_B,
                m,
                device_QT_even,
                device_QT_odd,
                device_QT_first,
                device_M_T,
                device_Σ_T,
                device_μ_Q,
                device_σ_Q,
                k,
                ignore_trivial,
                excl_zone,
                device_profile,
                device_indices,
                True,
            )
``` 

### STUMPY Installation

Technically, STUMPY can be `pip`/`conda` installed but since we are interested in injecting code, let's clone and install from source:

```
git clone https://github.com/TDAmeritrade/stumpy.git ./stumpy.git
cd stumpy.git
python -m pip install .
```

For GPUs, it is expected that the appropriate NVIDIA driver and cudatoolkit are also installed.

### STUMPY on GPUs

Below is a simple example using a small data set for running the `gpu_stump` function. `gpu_stump` is just a light wrapper that does a tiny bit of pre-work and then hands off the computation to `_compute_and_update_PI_kernel` GPU kernel:

```
import stumpy
import numpy as np

T = np.random.rand(10000)
m = 50

gpu_mp = stumpy.gpu_stump(T, m)
```

### Validating STUMPY GPU Results

Since we have both CPU and GPU implementations in STUMPY, we can validate that the GPU output is correct using our CPU code (which is appropriate for small to medium sized data that are less than 100K in length):

```
import stumpy
import numpy as np
import numpy.testing as npt

T = np.random.rand(10000)
m = 50

gpu_mp = stumpy.gpu_stump(T, m)
cpu_mp = stumpy.stump(T, m)

npt.assert_almost_equal(gpu_mp, cpu_mp)
```

**Comment by seanlaw** [comment]
@gmarkell Please let know if you need anything else or if there's anything I can explain/clarify! Thanks again for your help

**Comment by gmarkall** [comment]
@seanlaw Thanks for the writeup of steps! I'll take a look at this (but it's very close to the weekend here) so it might be next week before I come back to you.

**Comment by seanlaw** [comment]
No rush! Please enjoy your time

**Comment by gmarkall** [comment]
There is nothing preventing asynchronous launches for the kernel - the log shows a long stream of:

```
...
== CUDA [2056] DEBUG -- call driver api: cuCtxGetCurrent
== CUDA [2056] DEBUG -- call driver api: cuCtxGetDevice
== CUDA [2056] DEBUG -- call driver api: cuCtxGetCurrent
== CUDA [2056] DEBUG -- call driver api: cuCtxGetDevice
== CUDA [2056] DEBUG -- call driver api: cuLaunchKernel
== CUDA [2056] DEBUG -- call driver api: cuCtxGetCurrent
== CUDA [2057] DEBUG -- call driver api: cuCtxGetDevice
== CUDA [2057] DEBUG -- call driver api: cuCtxGetCurrent
== CUDA [2057] DEBUG -- call driver api: cuCtxGetDevice
== CUDA [2057] DEBUG -- call driver api: cuLaunchKernel
...
```

which is exactly what we'd hope for. One problem is that the kernels are quite short-lived, so the launch overhead dominates the execution time somewhat. I have a Numba PR that reduces the launch overhead (https://github.com/numba/numba/pull/6042), and timing the following (with Xeon Gold 6128, Quadro RTX 8000):

```python
import stumpy
import numpy as np
import numpy.testing as npt
import time

T = np.random.rand(10000)
m = 50

gpu_start = time.perf_counter()
gpu_mp = stumpy.gpu_stump(T, m)
gpu_end = time.perf_counter()

cpu_start = time.perf_counter()
cpu_mp = stumpy.stump(T, m)
cpu_end = time.perf_counter()

npt.assert_almost_equal(gpu_mp, cpu_mp)

print(f"CPU: {cpu_end - cpu_start}")
print(f"GPU: {gpu_end - gpu_start}")
```

With Numba master:

```
CPU: 6.643358725006692
GPU: 3.1323348870064365
```

With PR 6042:

```
CPU: 6.877583767010947
GPU: 1.2927859439951135
```

The execution time is reduced by 59% for this size input. When I change the size of T from 10000 to 100000, the GPU times are 31.97s before / 13.44s after. There is more that can be done in Numba to optimise the launch times, but that will (hopefully) be added to future PRs.

Another thought on this is that perhaps if the PR https://github.com/numba/numba/pull/4551 were picked up again so Numba supported cooperative groups, it would be possible to have a global barrier in the kernel that allows the `for i in range` loop from `_gpu_stump` to be pushed into the kernel, and then only have one launch in `_gpu_stump` - this would get rid of a lot of launch overhead, though it would have to be balanced by any limitation that it places on the size of the grid (hopefully not too much for this case).

**Comment by seanlaw** [comment]
> There is nothing preventing asynchronous launches for the kernel - the log shows a long stream of:

Maybe I'm a bit confused/misguided on the terminology here so please correct my understanding if I'm wrong. When you say "asynchronous launches" does that imply "asynchronous kernel execution"? I ask because we do need the `for i in range` loop to happen in order. So, the `i+1`th iteration MUST happen only after the `i`th iteration is completed. If the for loop happens in random order then the results should be wrong and not match the CPU output. The fact that the GPU results match the CPU results seems to imply that the kernels are executed in top down order according to the `for i in range` loop. Is that correct?

> One problem is that the kernels are quite short-lived, so the launch overhead dominates the execution time somewhat. I have a Numba PR that reduces the launch overhead (numba/numba#6042)

So, `gpu_stump` will typically be used when the data size is in the `T = np.random.rand(1_000_000)` (1 million) to `T = np.random.rand(100_000_000)` (100 million+) range. In our [very rough performance testing](https://github.com/TDAmeritrade/stumpy#performance), ~67 million data points takes around 13 days and 6 days on 1x 1080Ti and 2x 1080Ti, respectively. Being able to cut that down in half would be huge for us!

> Another thought on this is that perhaps if the PR numba/numba#4551 were picked up again so Numba supported cooperative groups, it would be possible to have a global barrier in the kernel that allows the for i in range loop from _gpu_stump to be pushed into the kernel, and then only have one launch in _gpu_stump - this would get rid of a lot of launch overhead, though it would have to be balanced by any limitation that it places on the size of the grid (hopefully not too much for this case).

I think being able to launch a single kernel and thereby reducing the kernel launch would be super! I have an [outstanding issue](https://github.com/TDAmeritrade/stumpy/issues/79) to write this up for Josh (Patterson) and Ben Zaitlen but I've been looking for a way to demonstrate bigger gains on GPUs compared with CPUs. Perhaps you would be interested in working on this together? I know you're super busy but it would be a really nice story to tell together.

**Comment by gmarkall** [comment]
> Maybe I'm a bit confused/misguided on the terminology here so please correct my understanding if I'm wrong. When you say "asynchronous launches" does that imply "asynchronous kernel execution"? I ask because we do need the `for i in range` loop to happen in order. So, the `i+1`th iteration MUST happen only after the `i`th iteration is completed. If the for loop happens in random order then the results should be wrong and not match the CPU output. The fact that the GPU results match the CPU results seems to imply that the kernels are executed in top down order according to the `for i in range` loop. Is that correct?

Asynchronous means that they are asynchronous with respect to the CPU, but not each other - they will be queued and executed in order on the default stream - so you are getting the behaviour you need. (To go back to your original Discourse question, streams are for when you don't mind if kernels are asynchronous with respect to each other - you can queue up two kernels in different streams if you don't mind which order they execute in)

(will respond to other two items later, just about to join a meeting)

**Comment by seanlaw** [comment]
> Asynchronous means that they are asynchronous with respect to the CPU, but not each other - they will be queued and executed in order on the default stream - so you are getting the behaviour you need. (To go back to your original Discourse question, streams are for when you don't mind if kernels are asynchronous with respect to each other - you can queue up two kernels in different streams if you don't mind which order they execute in)

Ahhhh! That makes perfect sense. I'm glad I asked the question. 😄  Thank you.

**Comment by gmarkall** [comment]
> So, `gpu_stump` will typically be used when the data size is in the `T = np.random.rand(1_000_000)` (1 million) to `T = np.random.rand(100_000_000)` (100 million+) range. In our [very rough performance testing](https://github.com/TDAmeritrade/stumpy#performance), ~67 million data points takes around 13 days and 6 days to compute on 1x 1080Ti and 2x 1080Ti, respectively. Being able to cut that down in half would be huge for us!

OK - I've done a quick try with N = 1_000_000, and the timings are 325 seconds before, and 135 seconds after - still about 59% faster. Are you able to benchmark on your hardware with PR #6042 to see whether you get similar results?

> I think being able to launch a single kernel and thereby reducing the kernel launch would be super! I have an [outstanding issue](https://github.com/TDAmeritrade/stumpy/issues/79) to write this up for Josh (Patterson) and Ben Zaitlen but I've been looking for a way to demonstrate bigger gains on GPUs compared with CPUs. Perhaps you would be interested in working on this together? I know you're super busy but it would be a really nice story to tell together.

I can try putting together something that allows testing this with a cooperative grid group - I'm a little unsure of the exact effort needed for this, and I'm also trying to wrap some things up this week due to being away next week, so I don't know whether it's something I'd get to this week. But I'm generally interested in trying to push on this.

**Comment by seanlaw** [comment]
> OK - I've done a quick try with N = 1_000_000, and the timings are 325 seconds before, and 135 seconds after - still about 59% faster. Are you able to benchmark on your hardware with PR #6042 to see whether you get similar results?

You'll have to excuse my ignorance here but should I be manually reinstalling `Numba` by cloning the commit from PR #6042? I've only ever installed `Numba` via `pip`/`conda` but never from source. Or do you think I could get away with replacing the three Python modules in my existing installation (i.e., inside of `site-packages`) with the three files that you had changed? 

> I can try putting together something that allows testing this with a cooperative grid group - I'm a little unsure of the exact effort needed for this, and I'm also trying to wrap some things up this week due to being away next week, so I don't know whether it's something I'd get to this week. But I'm generally interested in trying to push on this.

Absolutely no rush on this. I'm at your mercy! 😄  I already appreciate everything that you've done so far @gmarkall 🙏 

Out of curiosity, aside from `Numba` itself, is the "cooperative grid group" functionality limited to a particular version of the `cudatoolkit`? Currently, I'm guessing that most of my users are using something equal to or greater than version 10.1 (basically, people who are using tensorflow and PyTorch). I just want to make sure it's not a bleeding edge capability that would break backwards compatibility for people.

**Comment by gmarkall** [comment]
A note: I just came across, https://github.com/TDAmeritrade/stumpy/issues/64, which also suggests grid sync is required :-)

**Comment by gmarkall** [comment]
> Yes, I would be happy to benchmark it! You'll have to excuse my ignorance here but should I be manually reinstalling `Numba` by cloning the commit from PR [numba/numba#6042](https://github.com/numba/numba/pull/6042)? I've only ever installed `Numba` via `pip`/`conda` but never from source. Or do you think I could get away with replacing the three Python modules in my existing installation (i.e., inside of `site-packages`) with the three files that you had changed?

I would guess the cleanest way to do this is in a new conda environment:

```
conda create -n numbadev -c numba python=3.8 numpy scipy jinja2 cffi llvmlite
conda activate numbadev
git clone git@github.com:gmarkall/numba.git
cd numba
git checkout grm-native-kernel-launch
python setup.py develop
```

Then use your normal workflow for building / testing STUMPY. The Numba docs on getting set up for development are here, for reference: https://numba.readthedocs.io/en/stable/developer/contributing.html#getting-set-up

> > I can try putting together something that allows testing this with a cooperative grid group - I'm a little unsure of the exact effort needed for this, and I'm also trying to wrap some things up this week due to being away next week, so I don't know whether it's something I'd get to this week. But I'm generally interested in trying to push on this.
> 
> Absolutely no rush on this. I'm at your mercy! smile I already appreciate everything that you've done so far @gmarkall pray

I've tidied up the CG work and made a PR that is close to being ready: https://github.com/numba/numba/pull/6245

With this PR and a modified version of the kernel containing the loop over `i` and a grid sync, I get timings:

```
N         Time (seconds)  Original time (s)
========  ==============  =================
  10000:            0.08               3.13
 100000:            2.21              31.97
1000000:          147.37             325.00
```

There's some scope for improvement here - altering the block size makes a bit of difference to the timing as well.
 
> Out of curiosity, aside from `Numba` itself, is the "cooperative grid group" functionality limited to a particular version of the `cudatoolkit`? Currently, I'm guessing that most of my users are using something equal to or greater than version 10.1 (basically, people who are using tensorflow and PyTorch). I just want to make sure it's not a bleeding edge capability that would break backwards compatibility for people.

CGs are supported since CUDA 9.0, so that shouldn't be a problem. My understanding is that a Pascal or later (CC 6.0 or later) architecture is required for a cooperative grid launch.

**Comment by seanlaw** [comment]
> I would guess the cleanest way to do this is in a new conda environment:
> conda create -n numbadev -c numba python=3.8 numpy scipy jinja2 cffi llvmlite
conda activate numbadev
git clone git@github.com:gmarkall/numba.git
cd numba
git checkout grm-native-kernel-launch
python setup.py develop
Then use your normal workflow for building / testing STUMPY. The Numba docs on getting set up for development are here, for reference: https://numba.readthedocs.io/en/stable/developer/contributing.html#getting-set-up

Great! Thank you for the instructions. I will try this out and report back once I've computed the benchmark

> With this PR and a modified version of the kernel containing the loop over i and a grid sync, I get timings:

> N         Time (seconds)  Original time (s)
========  ==============  =================
  10000:            0.08               3.13
 100000:            2.21              31.97
1000000:          147.37             325.00
There's some scope for improvement here - altering the block size makes a bit of difference to the timing as well.

Hmmm, your previous timing (with only the changes from numba/numba#6042) showed ~59% improvement and this looks consistent with the `N = 1_000_000` result shown in the table directly above (which includes numba/numba#6042, a kernel with the for-loop, and grid sync). Do you think that we are approaching saturation and that the maximum percent increase will plateau around 59%? The results seem to suggest that, for `N >= 1_000_000`, pushing the for-loop into the kernel and adding grid sync is only helping marginally or, perhaps, not at all. Or maybe I'm interpreting this incorrectly?

> CGs are supported since CUDA 9.0, so that shouldn't be a problem. My understanding is that a Pascal or later (CC 6.0 or later) architecture is required for a cooperative grid launch.

That is very useful to know. Thank you!

**Comment by gmarkall** [comment]
> Hmmm, your previous timing (with only the changes from [numba/numba#6042](https://github.com/numba/numba/pull/6042)) showed ~59% improvement and this looks consistent with the `N = 1_000_000` result shown in the table directly above (which includes [numba/numba#6042](https://github.com/numba/numba/pull/6042), a kernel with the for-loop, and grid sync). Do you think that we are approaching saturation and that the maximum percent increase will plateau around 59%? The results seem to suggest that, for `N >= 1_000_000`, pushing the for-loop into the kernel and adding grid sync is only helping marginally or, perhaps, not at all. Or maybe I'm interpreting this incorrectly?

I agree with your interpretation - it looks like we plateau at 59% for reducing the launch overhead. I think the improvement that using CGs and having the loop inside the kernel makes is that it's much faster for smaller matrix sizes - e.g. N = 10000 - 100000.

**Comment by seanlaw** [comment]
@gmarkall I'm running some of the unit tests for STUMPY right now (before running the benchmark) and noticed that when I disable JIT and enable the CUDA simulator

```
export NUMBA_DISABLE_JIT=1
export NUMBA_ENABLE_CUDASIM=1
```

Then the gpu related tests (even though it isn't running on the GPU) are taking significantly longer than usual. So, I'm not sure if something needs to be addressed with the CUDA simulator within `Numba`. To be clear, I understand that the CUDA simulations will always take more time but we've been running these CUDA simulations for over a year now and this definitely anomalous. I've even compared it to another identical server with the same specs but without your version of Numba and it completes within the expected time.

**Comment by gmarkall** [comment]
What version of Numba are you using as the "reference"?

**Comment by seanlaw** [comment]
> What version of Numba are you using as the "reference"?

```
numba                     0.51.2           py37h9fdb41a_0    conda-forge
```

**Comment by seanlaw** [comment]
@gmarkall Here are my results using 1x 1080Ti 
```
# 1x 1080Ti, Numba 0.51.2 conda-forge
10000 5.3
100000 52.5
1000000 535.8

# 1x 1080Ti, Numba 0.51.2 gmarkall
10000 4.9
100000 49.2
1000000 490.9
```

and with 2x 1080Ti (on the same physical server)

```
# 2x 1080Ti, Numba 0.51.2 conda-forge
10000 4.8
100000 29.0
1000000 275.7

# 2x 1080Ti, Numba 0.51.2 gmarkall
10000 4.6
100000 26.7
1000000 254.1
```

Overall, there is a bit of performance gain using your `numba` version but not even close to the speeds that you had observed in your benchmark.

**Comment by seanlaw** [comment]
Okay, I think I spotted the problem. When I install `Numba` from your clone, `conda list | grep numba` shows:

```
numba                     0.15.1+13168.g526a66e34           dev_0    <develop>
```

However, when I install STUMPY, its `setup.py` requires `numba >= 0.42.0`. So, STUMPY actually ended up installing the newest version of `Numba` from `conda-forge` (I just tried a fresh re-install and noticed this) and that's likely why the numbers were about the same above. I've since corrected this and basically removed the `Numba` version dependency from STUMPY's `setup.py`. I'll report back with updated benchmarks as soon as I have them!

**Comment by seanlaw** [comment]
@gmarkall My apologies for the noise above. Here are the correct benchmark results:
 
```
# 1x 1080Ti, Numba 0.51.2 conda-forge
10000 5.3
100000 52.5
1000000 535.8

# 1x 1080Ti, Numba gmarkall
10000 2.1
100000 21.0
1000000 328.8
```

and with 2x 1080Ti (on the same physical server)

```
# 2x 1080Ti, Numba 0.51.2 conda-forge
10000 4.8
100000 29.0
1000000 275.7

# 2x 1080Ti, Numba gmarkall
10000 3.3
100000 13.6
1000000 169.2
```

Looks like it is consistent with your observations now!

**Comment by seanlaw** [comment]
Additionally, this is not unexpected but I also confirmed that the `gpu_stump` output produced from your Numba clone also match the CPU `stump` output.

**Comment by seanlaw** [comment]
@gmarkall This is probably expected but I ran some further benchmarks on larger data sets. 

With 2x 1080Ti (on the same physical server)

# 2x 1080Ti, Numba 0.51.2 conda-forge
10000 4.7
100000 29.1
16777216 48426.1
17729800 54259.8
33554432 194817.4

# 2x 1080Ti, Numba 0.51.2 gmarkall
10000 3.4
100000 13.0
16777216 47942.5
17729800 53529.8
33554432 192727.1

So, your version is certainly slightly faster (around 1%) for larger datasets.

**Comment by gmarkall** [comment]
https://github.com/numba/numba/pull/6042 is now merged in time for 0.52RC, so the speedups we observed earlier in this discussion should hopefully be reflected in the 0.52 release.

**Comment by gmarkall** [comment]
@seanlaw I guess this issue no longer needs to remain open?
